### ISSUE TITLE
[gemma4_e2b_q4nx] Prefill initialization exceeds the hardware-context limit

### ISSUE THREAD

**Issue by tonyjie** [Xilinx/mlir-air#2005]
Gemma-4-E2B prefill fails during weight preloading on this machine because it keeps **22 kernel hardware contexts resident**, exceeding the device's **16-context limit**. This happens before the decoder is constructed, at all four tested prefill lengths: 512, 1024, 2048 and 4096.

Related to the Gemma-4 investigation in #1984, but the context-allocation failure below is independently reproduced; a common root cause with that decode issue is not established.

**Environment**

Ryzen AI 9 HX 370 / Strix NPU2; Linux 6.17.0-14-generic; XRT 2.21; amdxdna DKMS source 7.0.0-rc1+git20260310.6b13cb8f4; Peano 22.0.0. Source checked at `deffe6f1`; existing kernels/compiler were built at `ac8c6436`. The relevant Gemma-4/cache code is unchanged between those revisions.

**Confirmed prefill failure**

`KernelCache.load_and_run()` in `programming_examples/llms/shared/infra/cache.py` retains a separate `XRTBackend` in `_loaded` for every kernel, without eviction. Gemma-4's `_preload()` reaches `up_w` as the seventeenth distinct kernel:

```text
DRM_IOCTL_AMDXDNA_CREATE_HWCTX IOCTL failed (err=-22): Invalid argument
```

AIR wraps this as failure to load `up_w.elf`. A weight-free probe using the existing prefill ELFs showed:

- All 22 kernels load successfully **individually**.
- Retaining contexts in preload order succeeds for the first 16 and fails on `up_w`.
- Releasing one existing context immediately lets the same `up_w` ELF load.

The driver source also specifies `hwctx_limit = 16` for this device. These are hardware execution contexts, separate from token context length.

A diagnostic wrapper capped prefill residency at 15, releasing least-recently-used backends while retaining weight/activation BOs. With unchanged ELFs and weights, all 35 layers then loaded at all four lengths. Original prompts matched the stored CPU **first token**, the Paris smoke prompt passed, and repeated prompts produced bit-identical logits. This does not yet validate full generated sequences.

Suggested fix: bounded context residency with buffer reuse and explicit headroom for the decoder.

**Separate decoder observation relevant to #1984**

With no prefill contexts loaded, upstream's synthetic `bench_decode.cpp` probe of the existing 35-layer decoder (capacity 1024, KV depth 512, geometry from `decode_geometry.py`) failed after ~64 seconds:

```text
qds_device::wait() unexpected command state
```

The process exited with SIGABRT before completing a timing batch. This is consistent with the failure class in #1984, but the exception did not expose an ERT state number; one attempt does not establish the same root cause or failure rate.

A subsequent NPU GELU smoke test passed all 131072 checked outputs. Drivers/runtimes were unchanged throughout.
